controllers/csidriver.go

```diff
@@ -0,0 +1,98 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1k8scsi "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/storage/v1"
+)
+
+type csiDriver struct {
+	name           string
+	fsGroupPolicy  string
+	attachRequired bool
+	mountInfo      bool
+	csiDriver      *v1k8scsi.CSIDriver
+	csiClient      v1.CSIDriverInterface
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (d csiDriver) createCSIDriverInfo(ctx context.Context) error {
+	// Create CSIDriver object
+	csiDriver := &v1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: d.name,
+		},
+		Spec: v1k8scsi.CSIDriverSpec{
+			AttachRequired: &d.attachRequired,
+			PodInfoOnMount: &d.mountInfo,
+		},
+	}
+	if d.fsGroupPolicy != "" {
+		policy := v1k8scsi.FSGroupPolicy(d.fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	driver, err := d.csiClient.Get(ctx, d.name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, err = d.csiClient.Create(ctx, csiDriver, metav1.CreateOptions{})
+		}
+		return err
+	}
+
+	// As FSGroupPolicy field is immutable, should be set only during create time.
+	// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+	if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+		return d.reCreateCSIDriverInfo(ctx)
+	}
+
+	// For csidriver we need to provide the resourceVersion when updating the object.
+	// From the docs (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata)
+	// > "This value MUST be treated as opaque by clients and passed unmodified back to the server"
+	csiDriver.ObjectMeta.ResourceVersion = driver.ObjectMeta.ResourceVersion
+	_, err = d.csiClient.Update(ctx, csiDriver, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d csiDriver) reCreateCSIDriverInfo(ctx context.Context) error {
+	err := d.csiClient.Delete(ctx, d.csiDriver.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete CSIDriver object for driver %q", d.csiDriver.Name)
+	}
+	_, err = d.csiClient.Create(ctx, d.csiDriver, metav1.CreateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to recreate CSIDriver object for driver %q", d.csiDriver.Name)
+	}
+	return nil
+}
+
+// deleteCSIDriverInfo deletes CSIDriverInfo and returns the error if any
+func (d csiDriver) deleteCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name string) error {
```

Comment:
This function is unused...? It's fine if not, I only see one place it could be used, and so I'd rather just get rid of this function.

Reply:
Yes, correct, it's unused for now we need it to uninstall csidrivers. i can remove it now and add it later.